agent/indy_catalyst_agent/messaging/credentials/manager.py

```diff
@@ -49,37 +51,90 @@ def context(self) -> RequestContext:
-        connection_id,
-        auto_issue=None,
-        credential_values=None,
+    async def prepare_send(
+        self, credential_definition_id, connection_id, credential_values
     ):
         """
         Create an offer.
 
         Args:
             credential_definition_id: Credential definition id for offer
             connection_id: Connection to create offer for
-            auto_issue: Should requests for this offer automatically be handled to
-                issue a credential
             credential_values: The credential values to use if auto_issue is enabled
 
         Returns:
-            A tuple (credential_exchange, credential_offer_message)
+            A tuple (
+                credential_exchange,
+                credential_offer_message
+            )
 
         """
 
-        # If auto_issue is False, ignore credential_values
-        if not auto_issue:
-            credential_values = None
+        cache: BaseCache = await self._context.inject(BaseCache)
+
+        # This cache is populated in credential_request_handler.py
+        # Do we have a source (parent) credential exchange for which
+        # we can re-use the credential request/offer?
+        source_credential_exchange_id = await cache.get(credential_definition_id)
+
+        if source_credential_exchange_id:
+
+            # Since we have the source exchange cache, we can re-use the schema_id,
+            # credential_offer, and credential_request to save a roundtrip
+            source_credential_exchange = await CredentialExchange.retrieve_by_id(
+                self._context, source_credential_exchange_id
+            )
 
-        # If auto_issue is True, credential_values must be provided
-        if auto_issue and not credential_values:
-            raise CredentialManagerError(
-                "auto_issue is True but credential_values is not set"
+            credential_exchange = CredentialExchange(
+                connection_id=connection_id,
+                initiator=CredentialExchange.INITIATOR_SELF,
+                credential_definition_id=credential_definition_id,
+                schema_id=source_credential_exchange.schema_id,
+                credential_offer=source_credential_exchange.credential_offer,
+                credential_request=source_credential_exchange.credential_request,
             )
+            await credential_exchange.save(self.context)
+
+            (
+                credential_exchange_record,
+                credential_message,
+            ) = await self.issue_credential(credential_exchange, credential_values)
+
+            # We use the source credential exchange's thread id as the parent thread id.
+            # This thread is a branch of that parent so that the other agent can use the
+            # parent thread id to look up its corresponding source credential exchange
+            # object as needed
+            thread = ThreadDecorator(pthid=source_credential_exchange.thread_id)
```

Comment:
fyi @cywolf this is where the pthid is used in this flow.

cc: @swcurran

Reply:
So we looked up the pre-existing credential offer and request and stored them on the new CredentialExchange record.. why does the other agent need to know that there's a link between the two? Because it expects to find a related credential offer as well?

Reply:
The other agent needs to lookup the corresponding “parent” exchange because it needs the credential metadata to be able to store it

Reply:
I looked up how agent-framework does it and it uses the same method to find the credential request metadata (from the thread ID, I don't think they store the request itself). As an opt-in optimization I think this method makes sense, although it would be nice if we could just find the metadata from the request automatically.